drivers: watchdog: Add support for the GD32 watchdog timers

soburi opened the pull request:
This PR adds watchdog drivers for GD32.
GD32 has two types of the watchdog timer, such as

- FWDGT free watchdog timer
- WWDGT Window watchdog timer

This PR adds support for both types of watchdogs.

I verified with  Longan Nano Board with GD32VF103.

More GD32 work #38657

soburi committed:
drivers: watchdog: Add GD32 Free watchdog timer driver

Add support for GD32 Free watchdog timer.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
drivers: watchdog: Add GD32 Window watchdog timer driver

Add support for GD32 Window watchdog timer.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
samples: driver: watchdog: Add GD32 SoCs support

Add support for boards that implement GD32 SoC.

Overlay file for longan_nano_lite enables FWDGT.

Add WDT_FEED_INTERVAL to make adjustable
the execution cycle of wdt_feed().

And also, make waiting for the WDT_FEED_INTERVAL period
for the watchdog window opens at the start of the test.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

martinjaeger commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -0,0 +1,41 @@
+config FWDGT_INITIAL_TIMEOUT
+	int "Value for FWDGT timeout in ms"
+	depends on FWDGT_GD32
+	default 1000
+	help
+	  Set timeout value for FWDGT in milliseconds.
+	  Not start FWDGT if set less than 0 value to this parameter.
+	  (same behavior as CONFIG_WDT_DISABLE_AT_BOOT set.)
+	  The minimum value is 1 ms.
+	  The following equation gives the maximum value.
+
+	    prescaler * reload / peripheral_freqency.
+
+	  Where, the prescaler = 256 and the reload = 4096.
+	  Each is the maximum value that can take.
+	  peripheral_frequency depends on SoC types.
+	  For example, This is 40000 for GD32VF103 and 32000 for GD32F4xx.
```
I suggest to limit the range via Kconfig and not allow negative values where the FWDGT will not start. If the watchdog should not start during boot, the appropriate Kconfig should be used instead.
```suggestion
	range 1 40000
	default 1000
	help
	  Set timeout value for FWDGT in milliseconds.
	  The following equation gives the maximum value:

	    prescaler * reload / peripheral_freqency.

	  where the maximum prescaler = 256 and the maximum reload = 4096.
	  The peripheral_frequency depends on the SoC type.
	  For example, this is 40000 for GD32VF103 and 32000 for GD32F4xx.
```

soburi replied:
FWDGT_GD32_INITIAL_TIMEOUT range changes in conjunction with IRC40K_VALUE.
GigaDevice's firmware design allows redefining IRC40K_VALUE by the user.

https://github.com/zephyrproject-rtos/hal_gigadevice/blob/242a7f4be7a3136606c7a65169775f7ef85ad444/gd32vf103/riscv/include/gd32vf103.h#L87

The maximum value of FWDGT_GD32_INITIAL_TIMEOUT defined as
   (256 * 4095 * 1000) / IRC40K_VALUE
But device-tree can use only numeric literals.
We can't define the range with that calculation in the device tree.

From the reason above, Not define the value range in the device tree and validating the FWDGT_GD32_INITIAL_TIMEOUT value range by preprocessor macros in a compilation.

martinjaeger commented on drivers/watchdog/wdt_fwdgt_gd32.c:
```diff
@@ -0,0 +1,161 @@
+	rcu_osci_on(DT_PROP(FWDGT_NODE, rcu_oscillator));
+	while (!rcu_osci_stab_wait(DT_PROP(FWDGT_NODE, rcu_oscillator))) {
+	}
+	/* TODO: It might better to reimplement with clock_control */
+	FWDGT_DATA(dev)->frequency = DT_PROP(FWDGT_NODE, frequency);
```
Is the problem that `clock_control` is not yet implemented for GD32? In general I prefer not to have code with `TODO` comments.

soburi replied:
I confirmed hardware specs.
Clock source configuration of the watchdogs are same in all GD32 series.
This line no need consideration for modifiable configuration.
I rewrite simply this line with using fixed value and remove TODO.

martinjaeger reviewed (changes requested):
Thanks @soburi for this contribution. Good to see the GD32 ecosystem support grow.

Got some rather small change requests and some questions below.

nandojve commented:
Hi @soburi ,

Rebase and squash update commits on the originals.

soburi commented:
> Hi @soburi ,
> 
> Rebase and squash update commits on the originals.

Hi @nandojve 
I was apply review comments with [Commit Suggestion] button.
I'm now working to correct this PR.
I will soon re-push.

soburi commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,189 @@
+#define WWDGT_COUNTER_MAX               (0x7fU)
+#define WWDGT_INTERNAL_DIVIDER          (4096ULL)
+
+#define WWDGT_DATA(dev) ((struct gd32_wwdgt_data *)(dev)->data)
```
I'll fix it that way.

soburi commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,189 @@
+ * @param timeout Timeout value in milliseconds.
+ * @param prescaler_exp exponent part of prescaler
+ *
+ * @return ticks count calculated by this formular.
```
I'll fix it that way.

soburi commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -0,0 +1,41 @@
+	help
+	  Enable the Free watchdog timer (FWDGT) driver for GD32 SoCs.
+
+config FWDGT_INITIAL_TIMEOUT
```
I'll fix it that way.

nandojve commented on modules/hal_gigadevice/Kconfig:
```diff
@@ -50,6 +50,13 @@ config GD32_HXTAL_25MHZ
 
+config GD32_LOW_SPEED_IRC_FREQUENCY
+	int "Low speed internal oscillator clock frequency"
+	default 32000 if SOC_SERIES_GD32F4XX
+	default 40000 if !SOC_SERIES_GD32F4XX
```
it can be good to use here too:

```c
	default 32000 if GD32_HAS_IRC_32K
	default 40000 if GD32_HAS_IRC_40K
```

nandojve commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,190 @@
+		return -ENOTSUP;
+	}
+
+	wwdgt_enable();
+	wwdgt_flag_clear();
+	wwdgt_interrupt_enable();
+	return 0;
```
```c
	wwdgt_enable();
	wwdgt_flag_clear();
	wwdgt_interrupt_enable();

	return 0;
```

nandojve commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,190 @@
+	data->callback = config->callback;
+	data->counter = counter;
+
+	wwdgt_config(counter, window, prescaler);
+	return 0;
```
```c
	wwdgt_config(counter, window, prescaler);

	return 0;
```

nandojve commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,190 @@
+{
+	struct gd32_wwdgt_data *data = dev->data;
+
+	ARG_UNUSED(channel_id);
+	wwdgt_counter_update(data->counter);
+	return 0;
```
```c
	ARG_UNUSED(channel_id);

	wwdgt_counter_update(data->counter);

	return 0;
```

nandojve commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,190 @@
+	struct gd32_wwdgt_data *data = dev->data;
+
+	if (wwdgt_flag_get()) {
+		wwdgt_flag_clear();
+		if (data->callback) {
+			data->callback(dev, 0);
+		}
```
```c
		wwdgt_flag_clear();

		if (data->callback) {
			data->callback(dev, 0);
		}
```

nandojve commented on samples/drivers/watchdog/src/main.c:
```diff
@@ -133,10 +148,14 @@ void main(void)
 
 	/* Feeding watchdog. */
 	printk("Feeding watchdog %d times\n", WDT_FEED_TRIES);
+#if WDT_MIN_WINDOW != 0
+	/* Wait open window. */
+	k_sleep(K_MSEC(WDG_FEED_INTERVAL));
+#endif
```
You are doing more than simple add overlay files. I think you should explain better at commit message why you are changing the example or split the changes.

soburi replied:
I got it. Splitting this commit.

nandojve reviewed (changes requested):
Hi @soburi ,

A few comments.

soburi commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -0,0 +1,28 @@
+	bool "GD32 Free watchdog timer (FWDGT) driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_GIGADEVICE_GD32_FWDGT))
+	select USE_GD32_FWDGT
+	select USE_GD32_DBG
```
USE_GD32_DBG must be enabled to use dbg_periph_enable. 

https://github.com/soburi/zephyr/blob/c4b8a0358d2c6b1b6dc1b9636a1f12890f413ced/drivers/watchdog/wdt_fwdgt_gd32.c#L77
https://github.com/soburi/zephyr/blob/c4b8a0358d2c6b1b6dc1b9636a1f12890f413ced/drivers/watchdog/wdt_wwdgt_gd32.c#L94


It is used to stop the watchdog counter at the CPU halted.

http://www.gd32mcu.com/data/documents/shujushouce/GD32VF103_User_Manual_EN_V1.2.pdf#page=1&zoom=auto,-218,842

nandojve replied:
Ok, Is this required in production? or is this only required when debugging board?

soburi replied:
Change to set the register directly not use the vendor provided function, and removing USE_GD32_DBG.

gmarull commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -0,0 +1,28 @@
+	help
+	  Enable the Free watchdog timer (FWDGT) driver for GD32 SoCs.
+
+config FWDGT_GD32_INITIAL_TIMEOUT
+	int "Value for FWDGT timeout in ms"
+	depends on FWDGT_GD32
+	default 1000
+	help
+	  Set timeout value for FWDGT in milliseconds.
+	  The following equation gives the maximum timeout value:
+
+	    timeout = prescaler * reload / peripheral_freqency.
+
+	  where the maximum prescaler = 256 and the maximum reload = 4096.
+	  The peripheral_frequency uses GD32_INTERNAL_LOW_SPEED_OSCILLATOR_FREQUENCY
+	  that defined in modules/hal_gigadevice/Kconfig.gd32.
+	  Validate the value is within a capable range at the compile time.
```
should be a DT parameter

soburi replied:
Remaking it as DT parameter.

gmarull commented on drivers/watchdog/wdt_fwdgt_gd32.c:
```diff
@@ -0,0 +1,163 @@
+{
+	int ret = 0;
+
+	/* Turn on and wait stabilize system clock oscillator. */
+	rcu_osci_on(RCU_SCK);
+	while (!rcu_osci_stab_wait(RCU_SCK)) {
+	}
```
should be handled by clock_control driver, see progress https://github.com/zephyrproject-rtos/zephyr/pull/41855

soburi replied:
OK, I'll try it.

gmarull commented on drivers/watchdog/wdt_fwdgt_gd32.c:
```diff
@@ -0,0 +1,163 @@
+{
+	ARG_UNUSED(dev);
+
+	if (options & WDT_OPT_PAUSE_HALTED_BY_DBG) {
+		dbg_periph_enable(DBG_FWDGT_HOLD);
+	}
```
`(options & ...) != 0U`

soburi replied:
Fixed it.

gmarull commented on drivers/watchdog/wdt_fwdgt_gd32.c:
```diff
@@ -0,0 +1,163 @@
+				      uint32_t *prescaler,
+				      uint32_t *reload)
+{
+	uint16_t divider = 4U;
+	uint8_t shift = 0U;
+	uint32_t ticks = (uint64_t)CONFIG_GD32_LOW_SPEED_IRC_FREQUENCY
+			 * timeout / MSEC_PER_SEC;
+
+	while ((ticks / divider) > FWDGT_RELOAD_MAX) {
+		shift++;
+		divider = 4U << shift;
+	}
+
+	if (!IS_VALID_FWDGT_PRESCALER(PSC_PSC(shift)) || timeout == 0) {
+		return -EINVAL;
+	}
+
+	/* convert the 'shift' to prescaler value */
+	*prescaler = PSC_PSC(shift);
+	*reload = (ticks / divider) - 1U;
```
nit: clock_control driver will also help here

gmarull commented on drivers/watchdog/wdt_fwdgt_gd32.c:
```diff
@@ -0,0 +1,163 @@
+#include <drivers/watchdog.h>
+#include <soc.h>
+
+#define FWDGT_NODE DT_NODELABEL(fwdgt)
```
not used

soburi replied:
Removed it.

gmarull commented on drivers/watchdog/wdt_fwdgt_gd32.c:
```diff
@@ -0,0 +1,163 @@
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT gd_gd32_free_watchdog
```
compatible should use GD terminology: fwdgt

soburi replied:
Fixed it.
free_watchdog -> fwdgt
window_watchdog -> wwdgt

gmarull reviewed (changes requested):
Please see my comments. I think these drivers would greatly benefit from having a proper clock control driver in place.

mbolivar-nordic reviewed (commented):
I think @nandojve and @gmarull are handling the DT portions of this PR very well, since they are both familiar with the process. So please consider me happy if/when they are happy :)

soburi committed:
tests: drivers: watchdog: Add support for GD32 SoCs

Enable watchdog tests for boards that implement GD32 SoC.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
I'd updated this PR and rebased to the newest.
@nandojve, 
I met a GitHub problem, so I failed to add comments to your mentions.
I already fixed it all.

martinjaeger commented:
FYI. The twister build failure is unrelated to this PR and I filed an issue already: https://github.com/zephyrproject-rtos/zephyr/issues/45611

soburi commented:
@martinjaeger 
> FYI. The twister build failure is unrelated to this PR and I filed an issue already: #45611

Thanks notifying me.
I've rebased it.

soburi commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -11,3 +12,12 @@ config FWDGT_GD32
+	bool "GD32 Window watchdog timer (WWDGT) Driver"
+	default $(dt_compat_enabled,$(DT_COMPAT_GIGADEVICE_GD32_WWDGT))
+	select USE_GD32_WWDGT
+	select USE_GD32_DBG
```
Update to remove USE_GD32_DBG selection.
(It is only used to set bits to registers. Replace by directly setting in code.)

soburi commented on modules/hal_gigadevice/Kconfig:
```diff
@@ -52,6 +52,23 @@ config GD32_HXTAL_25MHZ
+config GD32_HAS_IRC_40K
+	bool
+	default y if !SOC_SERIES_GD32F4XX
+	default n if SOC_SERIES_GD32F4XX
```
Renaming GD32_HAS_IRC_xxx to GD32_LOW_SPEED_IRC_xxx.
These make as items of GD32_LOW_SPEED_IRC_FREQUENCY choice.
(Following the style of GD32_HXTAL_FREQUENCY defined just before.

nandojve commented on modules/hal_gigadevice/Kconfig:
```diff
@@ -52,6 +52,35 @@ config GD32_HXTAL_25MHZ
 
 endchoice
 
+choice GD32_LOW_SPEED_IRC_FREQUENCY
+	prompt "Internal RC oscillator clock frequency"
+	default GD32_LOW_SPEED_IRC_32K if SOC_SERIES_GD32F4XX
+	default GD32_LOW_SPEED_IRC_40K if !SOC_SERIES_GD32F4XX
+	help
+	  Define value of low speed internal RC oscillator (IRC) in Hz
+	  This value sets the frequency of the oscillator.
+
+config GD32_LOW_SPEED_IRC_32K
+	bool "32KHz"
+	depends on SOC_SERIES_GD32F4XX
+	help
+	  Use 32KHz oscillator for low speed internal RC Oscillator
+
+config GD32_LOW_SPEED_IRC_40K
+	bool "40KHz"
+	depends on !SOC_SERIES_GD32F4XX
+	help
+	  Use 40KHz oscillator for low speed internal RC Oscillator
+
+endchoice
+
+config GD32_LOW_SPEED_IRC_HZ
+	int
+	default 32000 if GD32_LOW_SPEED_IRC_32K
+	default 40000 if GD32_LOW_SPEED_IRC_40K
+	help
+	  Numeric value of low speed IRC frequency (in HZ)
+
```
There is no need to define this choice, just keep below snip and adjust name at GD32_LOW_SPEED_IRC_HZ.

```
config GD32_HAS_IRC_32K
	bool
	help
	  Use 32KHz oscillator for low speed internal RC Oscillator

config GD32_HAS_IRC_40K
	bool
	help
	  Use 40KHz oscillator for low speed internal RC Oscillator
```

Then add on each SoC Kconfig.series the correct option. The **GD32_HAS_**_FOO_ should be defined on each SoC to keep Kconfig and CMake rules clean with more granularity.

```
 config SOC_SERIES_GD32F4XX 
 	bool "GigaDevice GD32F4XX series Cortex-M4F MCU" 
 	select ARM 
 	select CPU_HAS_ARM_MPU 
 	select CPU_HAS_FPU 
 	select CPU_CORTEX_M4 
 	select SOC_FAMILY_GD32_ARM 
 	select GD32_HAS_AF_PINMUX 
+ 	select GD32_HAS_IRC_32K
```

soburi replied:
Thanks for reviewing. 
And sorry for my late response.

```
+config GD32_DBG_SUPPORT
+	bool "Enable GD32 Debug features"
+	select USE_GD32_DBG
+	depends on DEBUG
# or even
+	default y if DEBUG
...
	bool "GD32 Free watchdog timer (FWDGT) driver"
	default $(dt_compat_enabled,$(DT_COMPAT_GIGADEVICE_GD32_FWDGT))
	select USE_GD32_FWDGT
-	select USE_GD32_DBG
```

I got it.
I apply your suggestion with using  `default y if DEBUG`

soburi replied:
> Then add on each SoC Kconfig.series the correct option. The GD32_HAS_FOO should be defined on each SoC to keep Kconfig and CMake rules clean with more granularity.

I understood it. And apply it.
By the way, I can only use GD32VF103 to verify this change.
I modified the source related to it.
Should I better apply this change to all other GD32-SoCs?

nandojve reviewed (changes requested):
HI @soburi ,

I share more comments because it is not clear to me if you understood my point. Sorry if I was vague.

> Update to remove USE_GD32_DBG selection.
(It is only used to set bits to registers. Replace by directly setting in code.)

The original code was OK to use it but I think I was not clear on my point, on my view we should define a way to enable debug features and this way Kconfig is clean and drivers will enable the feature only when user allows it.

At Kconfig:
```diff
+config GD32_DBG_SUPPORT
+	bool "Enable GD32 Debug features"
+	select USE_GD32_DBG
+	depends on DEBUG
# or even
+	default y if DEBUG
...
	bool "GD32 Free watchdog timer (FWDGT) driver"
	default $(dt_compat_enabled,$(DT_COMPAT_GIGADEVICE_GD32_FWDGT))
	select USE_GD32_FWDGT
-	select USE_GD32_DBG
```

At Drivers:
```diff
	if ((options & WDT_OPT_PAUSE_HALTED_BY_DBG) != 0U) {
+#if GD32_DBG_SUPPORT
		dbg_periph_enable(DBG_FWDGT_HOLD);
+#else
+		LOG_ERR("Pause when halted by debugger not supported, see GD32_DBG_SUPPORT");
+		return -ENOTSUP;
+#endif
	}
```

martinjaeger commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,199 @@
+
+static void gd32_wwdgt_irq_config(const struct device *dev);
+
+static inline struct gd32_wwdgt_data *get_dev_data(const struct device *dev)
+{
+	return dev->data;
+}
```
This function is not used and should be removed.

soburi replied:
Sorry for my late response.
I removed it.

soburi committed:
modules: Add IRC configurations

Add options about Internal RC(IRC) oscillator.

- GD32_HAS_IRC_32K/40K indicates IRC types.
- GD32_LOW_SPEED_IRC_FREQUENCY is the numeric value of frequency

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
modules: Add option for enabling HAL debug functions

Add GD32_DBG_SUPPORT options to enable HAL debug functions.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

fkokosinski commented:
Hi @soburi, thanks for your contribution. Could you rebase your changes, as there are conflicting files right now?

soburi committed:
boards: Enable watchdog node for boards that implements GD32 SoCs

Enable watchdog in each GD32 implemented board.
Add the watchdog function as a supported function in each yaml files.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi committed:
dts: Add watchdog configuration for GD32 SoCs

Add watchdog node for GD32 series SoC's dts files.

Signed-off-by: TOKITA Hiroshi <tokita.hiroshi@gmail.com>

soburi commented:
Hi, @fkokosinski

Rebase done.

CI failure is not related this PR.
Update tests/drivers/watchdog and samples/drivers/watchdog to apply #48030 #47912 modifications.

soburi commented:
> CI failure is not related this PR.

It should fix with #48393.

nandojve commented on modules/hal_gigadevice/Kconfig:
```diff
@@ -69,6 +69,11 @@ config GD32_LOW_SPEED_IRC_FREQUENCY
+config GD32_DBG_SUPPORT
+	bool "Use GD32 Debug features"
+	select USE_GD32_DBG
+	default y if DEBUG
+
```
I was thinking more about this option. You can switch to:

```diff
-default y if DEBUG
+default y
```

soburi replied:
OK. Applied it.

nandojve commented on samples/drivers/watchdog/boards/gd32e103v_eval.conf:
```diff
@@ -0,0 +1,4 @@
+# Copyright (c) 2022 TOKITA Hiroshi <tokita.hiroshi@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_GD32_DBG_SUPPORT=y
```
Since we relaxed this condition condition, you can drop all this *.conf

soburi replied:
Remove all *.conf

soburi commented on tests/drivers/watchdog/wdt_basic_api/testcase.yaml:
```diff
@@ -30,3 +31,17 @@ tests:
+    platform_allow: gd32f450i_eval gd32e103v_eval gd32f450z_eval gd32f470i_eval
+      gd32f350r_eval gd32f403z_eval longan_nano gd32vf103v_eval gd32vf103c_starter
+  drivers.watchdog.gd32wwdgt:
+    filter: dt_compat_enabled("gd,gd32-fwdgt") or dt_compat_enabled("gd,gd32-fwdgt")
```
Same as above

soburi commented on tests/drivers/watchdog/wdt_basic_api/testcase.yaml:
```diff
@@ -30,3 +31,17 @@ tests:
     extra_args: OVERLAY_CONFIG="boards/nrf52840dk_nrf52840_counter.conf"
       DTC_OVERLAY_FILE="boards/nrf52840dk_nrf52840_counter.overlay"
+  drivers.watchdog.gd32fwdgt:
+    filter: dt_compat_enabled("gd,gd32-fwdgt") or dt_compat_enabled("gd,gd32-fwdgt")
```
Correcting to 
```
    filter: dt_compat_enabled("gd,gd32-fwdgt") or dt_compat_enabled("gd,gd32-wwdgt")
```
and rework some configuration to follow current source

soburi commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -10,3 +11,11 @@ config FWDGT_GD32
+	select USE_GD32_WWDGT
+	depends on (SOC_FAMILY_GD32 || SOC_SERIES_GD32VF103)
+	help
+	  Enable the Window watchdog timer (WWDGT) driver for GD32 SoCs.
```
Fixed it

soburi commented on drivers/watchdog/wdt_fwdgt_gd32.c:
```diff
@@ -0,0 +1,179 @@
+#else
+#define RCU_IRC_LOW_SPEED RCU_IRC40K
+#endif
+
```
Apply it.

soburi commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -0,0 +1,12 @@
+	select USE_GD32_FWDGT
+	depends on (SOC_FAMILY_GD32 || SOC_SERIES_GD32VF103)
+	help
+	  Enable the Free watchdog timer (FWDGT) driver for GD32 SoCs.
```
Fixed it

soburi commented on modules/hal_gigadevice/CMakeLists.txt:
```diff
@@ -30,6 +30,14 @@ elseif(${CONFIG_GD32_HXTAL_25MHZ})
 endif()
 
+if(CONFIG_GD32_HAS_IRC_32K)
+  zephyr_compile_definitions(IRC32K_VALUE=${CONFIG_GD32_HAS_IRC_32K})
```
Fixed it.

soburi commented on modules/hal_gigadevice/CMakeLists.txt:
```diff
@@ -30,6 +30,14 @@ elseif(${CONFIG_GD32_HXTAL_25MHZ})
+endif()
+
+if(CONFIG_GD32_HAS_IRC_40K)
+  zephyr_compile_definitions(IRC32K_VALUE=${CONFIG_GD32_HAS_IRC_40K})
```
Fixed it.

carlocaione commented on modules/hal_gigadevice/Kconfig:
```diff
@@ -70,6 +70,11 @@ config GD32_LOW_SPEED_IRC_FREQUENCY
 	help
 	  Define value of low speed internal RC oscillator (IRC) in Hz
 
+config GD32_DBG_SUPPORT
+	bool "Use GD32 Debug features"
+	select USE_GD32_DBG
+	default y
```
Isn't it possible to directly use `USE_GDB32_DBG`. This new symbol seems redundant to me.

soburi replied:
@carlocaione, @nandojve 

It can, but should better keep it to allow to turn off the debug function, I think.
A user is usually not allowed to configure the USE_... option in the modules/hal_* directory. (It seems to be an implicit rule.)
Adding GD32_DBG_SUPPORT is to enable/disable the debug function by the user.

nandojve replied:
The _USE_foo_ and _HAS_foo_ symbols are not visible. The _USE_foo_ have the meaning to select something to build inside hal itself. I think there is no intention to have that kind of symbol inside zephyr mainline **code**. So, it seems to be redundant because, in general, zephyr drivers have been selecting the _USE_foo_ feature. In this case, GD32_DBG_SUPPORT is used to enable the hal_feature itself and can be used inside zephyr code. Unfortunately it is not possible to simple rely on the global DEBUG symbol.

@soburi, maybe it is important add documentation using `help` description for this config once it is used inside zephyr's code.

soburi replied:
@nandojve ,
Added `help` description about for this config.

carlocaione commented on dts/bindings/watchdog/gd,gd32-fwdgt.yaml:
```diff
@@ -0,0 +1,26 @@
+
+    initial-timeout:
+      type: int
+      default: 1000
+      description: |
```
Please document the default value

nandojve commented:
@gmarull, @carlocaione ,
I was wondering if there are anything missing from your side.

soburi commented:
Sorry, wrong commit pushed.
Already restored.

fabiobaltieri commented:
@gmarull @martinjaeger can you re-review this? (you both have a requested changes pending)

gmarull commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -0,0 +1,13 @@
+# Copyright (c) 2021 TOKITA Hiroshi <tokita.hiroshi@gmail.com>
+# SPDX-License-Identifier: Apache-2.0
+
+DT_COMPAT_GIGADEVICE_GD32_FWDGT := gd,gd32-fwdgt
```
this is redundant

soburi replied:
Removed it.

gmarull commented on drivers/watchdog/wdt_fwdgt_gd32.c:
```diff
@@ -0,0 +1,178 @@
+#include <zephyr/drivers/watchdog.h>
+
+#include <gd32_fwdgt.h>
+#include <soc.h>
```
not needed

soburi replied:
Removed this line.

gmarull commented on dts/bindings/watchdog/gd,gd32-fwdgt.yaml:
```diff
@@ -0,0 +1,27 @@
+    reg:
+      required: true
+
+    initial-timeout:
```
use `-ms` suffix as property has units.

soburi replied:
Added it.

gmarull commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -2,6 +2,7 @@
 # SPDX-License-Identifier: Apache-2.0
 
 DT_COMPAT_GIGADEVICE_GD32_FWDGT := gd,gd32-fwdgt
+DT_COMPAT_GIGADEVICE_GD32_WWDGT := gd,gd32-wwdgt
```
ditto

soburi replied:
Removed this line.

gmarull commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,205 @@
+#include <zephyr/drivers/watchdog.h>
+
+#include <gd32_wwdgt.h>
+#include <soc.h>
```
ditto

soburi replied:
Removed this line.

gmarull commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,205 @@
+
+#define DT_DRV_COMPAT gd_gd32_wwdgt
+
+#include <zephyr/drivers/watchdog.h>
+
+#include <gd32_wwdgt.h>
+#include <soc.h>
+
+#define LOG_LEVEL CONFIG_WDT_LOG_LEVEL
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(wdt_wwdgt_gd32);
```
```c
#include <zephyr/drivers/watchdog.h>
#include <zephyr/logging/log.h>

#include <gd32_wwdgt.h>

LOG_MODULE_REGISTER(wdt_wwdgt_gd32, CONFIG_WDT_LOG_LEVEL);
```

soburi replied:
Corrected as this pointing out.

gmarull commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,205 @@
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(wdt_wwdgt_gd32);
+
+#define WWDGT_NODE DT_NODELABEL(wwdgt)
```
why is this needed? you have `DT_INST*` available

soburi replied:
Use DT_INST_PROP(0, ...)

gmarull commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,205 @@
+
+static void gd32_wwdgt_irq_config(const struct device *dev);
+
+/**
+ * @param timeout Timeout value in milliseconds.
+ * @param exp exponent part of prescaler
+ *
+ * @return ticks count calculated by this formula.
+ *
+ *  timeout = pclk * INTERNAL_DIVIDER * (2^prescaler_exp) * (count + 1)
+ *  transform as
+ *  count = (timeout * pclk / INTERNAL_DIVIDER * (2^prescaler_exp) ) - 1
+ *
+ *  and add WWDGT_COUNTER_MIN to this as a offset value.
+ */
+static inline uint32_t gd32_wwdgt_calc_ticks(uint32_t timeout, uint32_t exp)
+{
+	const uint32_t pclk = rcu_clock_freq_get(CK_APB1);
+
+	return ((timeout * pclk)
+		/ (WWDGT_INTERNAL_DIVIDER * (1 << exp) * MSEC_PER_SEC) - 1)
+		+ WWDGT_COUNTER_MIN;
+}
```
please use clock driver

soburi replied:
Query the clock frequency with clock_control_get_rate here.

gmarull commented on drivers/watchdog/wdt_wwdgt_gd32.c:
```diff
@@ -0,0 +1,205 @@
+
+static int gd32_wwdgt_init(const struct device *dev)
+{
+	rcu_periph_clock_enable(DT_PROP(WWDGT_NODE, rcu_periph_clock));
```
use `clocks` property, also reset peripheral using reset API.

soburi replied:
Replaced with clock_control_on()

gmarull commented on dts/riscv/gigadevice/gd32vf103.dtsi:
```diff
@@ -183,6 +183,21 @@
+			compatible = "gd,gd32-wwdgt";
+			reg = <0x40002C00 0x400>;
+			rcu-periph-clock = <0x70b>;
+			interrupt-parent = <&eclic>;
```
redundant, parent is set at `soc` level now

soburi replied:
Removed this line.

gmarull commented on samples/drivers/watchdog/sample.yaml:
```diff
@@ -27,3 +27,15 @@ tests:
+  sample.drivers.watchdog.gd32_wwdgt:
+    extra_args: DTC_OVERLAY_FILE=boards/gd32_wwdgt.overlay
+    filter: dt_compat_enabled("gd,gd32-wwdgt")
+    platform_allow: gd32e103v_eval gd32e507v_start gd32f350r_eval gd32f403z_eval
+        gd32f450i_eval gd32f450z_eval gd32f470i_eval gd32vf103c_starter
+        gd32vf103v_eval longan_nano
```
this seems redundant with the filter, same above.

soburi replied:
Rewrite with filter: property

soburi replied:
This test gives the DTC_OVERLAY_FILE option. It causes compile failure on other platforms.
This `platform_allow` line is needed.

soburi commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -0,0 +1,13 @@
+	bool "GD32 Free watchdog timer (FWDGT) driver"
+	default y
+	depends on DT_HAS_GD_GD32_FWDGT_ENABLED
+	depends on (SOC_FAMILY_GD32 || SOC_SERIES_GD32VF103)
```
Removed it.

soburi commented on drivers/watchdog/wdt_fwdgt_gd32.c:
```diff
@@ -0,0 +1,178 @@
+	ARG_UNUSED(dev);
+
+	if ((options & WDT_OPT_PAUSE_HALTED_BY_DBG) != 0U) {
+#if IS_ENABLED(CONFIG_GD32_DBG_SUPPORT)
```
Corrected as this pointing out.

soburi commented on drivers/watchdog/Kconfig.gd32:
```diff
@@ -11,3 +12,12 @@ config FWDGT_GD32
+	bool "GD32 Window watchdog timer (WWDGT) Driver"
+	default y
+	depends on DT_HAS_GD_GD32_WWDGT_ENABLED
+	depends on (SOC_FAMILY_GD32 || SOC_SERIES_GD32VF103)
```
Removed this line.

mmahadevan108 commented:
@soburi. I have cleared the 3.2 milestone as we are past the feature freeze date. Kindly pick a new milestone.

soburi commented:
Hi, @mmahadevan108 

Thank you for the notification.
I think this PR will merge soon.
I will set a target release at that time.

soburi commented on dts/bindings/watchdog/gd,gd32-fwdgt.yaml:
```diff
@@ -0,0 +1,27 @@
+      default: 100
+      description: |
+        Set timeout value in milliseconds.
+        The default value is set as 0.1sec(100msec).
```
Reconsidered the default value.
And added a description of why the value.

fkokosinski commented:
Hi @soburi, you seem to have introduced undefined references to `log_output_flush` in your code and some minor style issues, which is why the CI now fails.

soburi commented:
Hi, @fkokosinski 
> Hi @soburi, you seem to have introduced undefined references to `log_output_flush` in your code and some minor style issues, which is why the CI now fails.

Rebase and fix CI failure done.

gmarull commented on dts/bindings/watchdog/gd,gd32-wwdgt.yaml:
```diff
@@ -0,0 +1,15 @@
+    reg:
+      required: true
+
+    clocks:
+      required: true
```
nit: resets required true missing

soburi replied:
It is significant issue!
Fixed it.

gmarull reviewed (commented):
one last nitpick, thanks for this work

nandojve reviewed (approved):
Thank you @soburi !
